xarray/core/dataset.py

```diff
@@ -1955,7 +1955,35 @@ def fillna(self, value):
         -------
         Dataset
         """
-        out = self._fillna(value)
+        if utils.is_dict_like(value):
+            value_keys = value.data_vars.keys() if isinstance(value, Dataset)\
+                else value.keys()
+            if not set(value_keys) <= set(self.data_vars.keys()):
+                raise ValueError('all variables in the argument to `fillna` '
+                                 'must be contained in the original dataset')
+        out = ops.fillna(self, value, join="left")
+        out._copy_attrs_from(self)
+        return out
```

Comment:
This would be a good time to remove all the `fillna` specific logic from `Dataset._calculate_binary_op` that is no longer used